Add glossary to documentation (#3352)
pydata/xarray: doc/terminology.rst

.. _terminology:

.. https://github.com/pydata/xarray/issues/2410

> .. https://github.com/pydata/xarray/issues/1295

Reviewer: Forgive me for my lack of docs understanding—what do these do?

Author: These comments were just for myself: issues where Xarray contributors or maintainers defined words. Removing them now.

> Terminology

Reviewer 2: Glossary?

Author: Maybe? When I hear "glossary", I think of an alphabetized resource that I can ignore until I need it. This is a slightly more opinionated glossary designed to introduce terms in a specific order to new users. That said, I don't have a strong opinion and don't know if "terminology" is any better.

Reviewer 2: :+1:

===========

> *Xarray terminology differs slightly from CF and mathematical conventions, and therefore using xarray, understanding the documentation, and parsing error messages is easier once key terminology is defined. This glossary was designed so that more fundamental concepts come first. Thus for new users, this page is best read top-to-bottom. Throughout the glossary,* ``arr`` *will refer to an xarray* :py:class:`DataArray` *in any small examples. For more complete examples, please consult the relevant documentation.*

Reviewer: ```suggestion
*Xarray terminology differs slightly from CF, mathematical conventions, and pandas; and therefore using xarray, understanding the documentation, and parsing error messages is easier once key terminology is defined. This glossary was designed so that more fundamental concepts come first. Thus for new users, this page is best read top-to-bottom. Throughout the glossary,* ``arr`` *will refer to an xarray* :py:class:`DataArray` *in any small examples. For more complete examples, please consult the relevant documentation.*
```

> ----

Reviewer 2: can you add a screenshot of how it looks?

Author: Yeah, will do once I've gone through the current feedback, integrated it, and rebuilt.

> **DataArray:** A multi-dimensional array with labeled or named dimensions. If its optional ``name`` property is set, it is a *named DataArray*.

Reviewer 2: Add something like "DataArrays add metadata — dimension names, coordinates, and attributes (see below) — to underlying "unlabeled" data structures such as numpy arrays and Dask arrays."?

Author: This is great. Added.

----

> **Dimension / dimensions:** A *dimension* is a nonnegative number for the dimensionality of the underlying data, while an array's *dimensions* are a set of dimension names. The name of the ``i``-th dimension is ``arr.dims[i]``. If an array is created without dimensions, the default dimension names are ``dim_0``, ``dim_1``, and so forth.

Reviewer: I'm not sure about `A *dimension* is a nonnegative number for the dimensionality `—I think that's a good definition for dimensionality, but wouldn't think of a dimension as a number?

Author: I struggled with this because in math, a "dimension" is a number, but in common parlance and xarray, "dimension" is often used to refer to a _dimension axis_—or a locus in which all but one of the coordinates ("coordinates" in the math sense, not xarray sense) is fixed. I didn't want to define "coordinate" and "dimension" in the math sense and then re-define them, although I think that would be the clearest at the expense of brevity.

Anyway, how about this:

> **Dimension:** In mathematics, the *dimension* of data is loosely the number of degrees of freedom for it. A *dimension axis* is a set of all points in which all but one of these degrees of freedom is fixed. We can think of each dimension axis as having a name, for example the "x dimension".  In xarray, a `DataArray`'s *dimensions* are its named dimension axes, and the name of the ``i``-th dimension is ``arr.dims[i]``. If an array is created without dimensions, the default dimension names are ``dim_0``, ``dim_1``, and so forth.

Reviewer 3: It might be worth borrowing the description from the netCDF data model documentation: https://www.unidata.ucar.edu/software/netcdf/docs/netcdf_data_model.html

> Dimensions: describe the axes of the [DataArrays]. A dimension has a name and a length...

----

> **Coordinate:** A one-dimensional array that labels a dimension of another ``DataArray``. There are two types of coordinate arrays: *dimension coordinates* and *non-dimension coordinates* (see below). A coordinate named ``x`` can be retrieved from ``arr.coords[x]``. A ``DataArray`` can have more coordinates than dimensions because a single dimension can be assigned multiple coordinate arrays. However, only one coordinate array can be a assigned as a particular dimension's dimension coordinate    array. As a consequence, ``len(arr.dims) <= len(arr.coords)`` in general.

Reviewer: non-dimensional coordinates can be multi-dimensional

Author: I thought "multi-dimensional coordinates" really just meant multiple one-dimensional coordinate arrays labeling the same dimension. I inferred that from [this page](https://xarray.pydata.org/en/stable/examples/multidimensional-coords.html). Is this what you mean or something else? Also, could you please show me an example? I've tried and failed to use `assign_coords` to add a multi-dimensional `DataArray` as a coordinate array.

Reviewer: For example

```python

In [8]: xr.DataArray(np.ones((2,3)), dims=['a','b'], coords=dict(c=(('a','b'), np.zeros((2,3)))))
Out[8]:
<xarray.DataArray (a: 2, b: 3)>
array([[1., 1., 1.],
       [1., 1., 1.]])
Coordinates:
    c        (a, b) float64 0.0 0.0 0.0 0.0 0.0 0.0
Dimensions without coordinates: a, b
```

> **Coordinate:** A one-dimensional array that labels a dimension of another ``DataArray``. There are two types of coordinate arrays: *dimension coordinates* and *non-dimension coordinates* (see below). A coordinate named ``x`` can be retrieved from ``arr.coords[x]``. A ``DataArray`` can have more coordinates than dimensions because a single dimension can be assigned multiple coordinate arrays. However, only one coordinate array can be a assigned as a particular dimension's dimension coordinate    array. As a consequence, ``len(arr.dims) <= len(arr.coords)`` in general.

Reviewer 2: "Coordinate values can be thought of as tick labels along a dimension"?

Author: Added!

.. note::

    The :py:class:`Variable` class is low-level interface and can typically be ignored. However, the word "variable" appears often enough in the code and documentation that is useful to understand.

----

> **Dimension:** In mathematics, the *dimension* of data is loosely the number of degrees of freedom for it. A *dimension axis* is a set of all points in which all but one of these degrees of freedom is fixed. We can think of each dimension axis as having a name, for example the "x dimension".  In xarray, a ``DataArray`` object's *dimensions* are its named dimension axes, and the name of the ``i``-th dimension is ``arr.dims[i]``. If an array is created without dimensions, the default dimension names are ``dim_0``, ``dim_1``, and so forth.

Reviewer: Well done on improving this part. It is fairly abstract now! We could add some examples, e.g.

> For example, the dimensions of an array of temperatures could be latitude, longitude, height, and date.

(ok in another PR though)

.. note::

>     **Name matching rules:** Xarray follows simple but important-to-grok name matching rules for dimensions and coordinates. Let ``arr`` be an array with an existing dimension ``x`` and assigned new coordinates ``new_coords``. If ``new_coords`` is a list-like collection, then they must be assigned a name that matches an existing dimension. For example, if ``arr.assign_coords({'x': new_coords}).``

Reviewer 2: ```suggestion
    **Name matching rules:** Xarray follows simple but important-to-understand name matching rules for dimensions and coordinates. Let ``arr`` be an array with an existing dimension ``x`` and assigned new coordinates ``new_coords``. If ``new_coords`` is a list-like for e.g. ``[1, 2, 3]`` then they must be assigned a name that matches an existing dimension. For example, if ``arr.assign_coords({'x': [1, 2, 3]}).``
```

> **Coordinate:** An array that labels a dimension of another ``DataArray``. Loosely, the coordinate array's values can be thought of as tick labels along a dimension. There are two types of coordinate arrays: *dimension coordinates* and *non-dimension coordinates* (see below). A coordinate named ``x`` can be retrieved from ``arr.coords[x]``. A ``DataArray`` can have more coordinates than dimensions because a single dimension can be assigned multiple coordinate arrays. However, only one coordinate array can be a assigned as a particular dimension's dimension coordinate    array. As a consequence, ``len(arr.dims) <= len(arr.coords)`` in general.

Reviewer: ```suggestion
**Coordinate:** An array that labels a dimension of another ``DataArray``. Loosely, the coordinate array's values can be thought of as tick labels along a dimension. There are two types of coordinate arrays: *dimension coordinates* and *non-dimension coordinates* (see below). A coordinate named ``x`` can be retrieved from ``arr.coords[x]``. A ``DataArray`` can have more coordinates than dimensions because a single dimension can be assigned multiple coordinate arrays. However, only one coordinate array can be a assigned as a particular dimension's dimension coordinate array. As a consequence, ``len(arr.dims) <= len(arr.coords)`` in general.
```

>     However, if ``new_coords`` is a one-dimensional ``DataArray``, then the rules are slightly more complex. In this case, if both ``new_coords``'s name and only dimension match any dimension name in ``arr.dims``, it is assigned as a dimension coordinate to ``arr``. If ``new_coords``'s name matches a name in ``arr.dims`` but its own dimension name does not, it is assigned as a non-dimension coordinate with name ``new_coords.dims[0]`` to ``arr``. Otherwise, an exception is raised.

Reviewer 2: Maybe link to the doc page instead of describing this here? Something like, if `new_coords` is a one-dimensional ``DataArray``, the rules can be more complex, see here: ...

Author: Ah, good idea. Is there somewhere in the docs that the name matching rules are fully described that I can link to?

----

> **Index:** An *index* is a :py:class:`pandas.Index` that indexes the values in a dimension coordinate. Non-dimension coordinates are not indexed. The index associated with dimension name ``x`` can be retrieved by ``arr.indexes[x]``. By construction, ``len(arr.dims) == len(arr.indexes)``

Reviewer 2: hahaha "An index is an Index that indexes..." :)

I don't know how to describe this better. 

Pandas says "Immutable ndarray implementing an ordered, sliceable set. The basic object storing axis labels for all pandas objects." 

The basic object bit is nice, though what is the difference between a dimension coordinate and an index?

Author: Ha, that's a fairly recursive definition. How about:

> **Index:** An *index* is a data structure optimized for efficient selecting and slicing of an associated array. Xarray creates indexes for dimension coordinates so that operations along dimensions are fast, while non-dimension coordinates are not indexed. Under the hood, indexes are implemented as `pandas.Index` objects. The index associated with dimension name ``x`` can be retrieved by ``arr.indexes[x]``. By construction, ``len(arr.dims) == len(arr.indexes)``

Reviewer: I _really_ like the start. I think this glossary should stay at the conceptual level as much as possible, rather than going into the implementation / pandas / etc. Saying it's optimized for selecting etc is fine, since it helps explain the concepts

----

> **Dataset:** A dict-like collection of ``DataArray`` objects with aligned dimensions. Thus, most operations that can be performed on the dimensions of a single ``DataArray`` can be performed on a dataset.

Reviewer 2: Move Dataset all the way up top? 

"Datasets have data variables (data_vars), dimensions (dims), coordinates (coords) and attributes"

Author: Moved it to right after `DataArray` and added a version of that sentence. I also moved **Variable** up so the order is: `DataArray`, `Dataset`, `Variable`, `dimension`, ... Otherwise, I think "data variables" is confusing. LMK if that sounds good or if you liked `Variable` near the bottom since it's a low-level implementation detail to new users.

Reviewer 2: I think this is fine.

    **Name matching rules:** Xarray follows simple but important-to-understand name matching rules for dimensions and coordinates. Let ``arr`` be an array with an existing dimension ``x`` and assigned new coordinates ``new_coords``. If ``new_coords`` is a list-like for e.g. ``[1, 2, 3]`` then they must be assigned a name that matches an existing dimension. For example, if ``arr.assign_coords({'x': [1, 2, 3]}).``

>     However, if ``new_coords`` is a one-dimensional ``DataArray``, then the rules are slightly more complex. In this case, if both ``new_coords``'s name and only dimension match any dimension name in ``arr.dims``, it is assigned as a dimension coordinate to ``arr``. If ``new_coords``'s name matches a name in ``arr.dims`` but its own dimension name does not, it is assigned as a non-dimension coordinate with name ``new_coords.dims[0]`` to ``arr``. Otherwise, an exception is raised.

Reviewer: Feels overly complicated but I think also just reality. I could see us eliding this here with a link to somewhere in the docs, given it's more about 'how-to' than terminology. Not a strong view though

Reviewer 2: Yeah, lets just leave this for another page in the docs.

Author: Okay, I'll remove this for now, and we can add these name matching rules later.
